lib/bots/keyframebot.ts

```diff
@@ -0,0 +1,684 @@
+        const headerToken = req.get('Authorization') || '';
+        const productSplitRepo = this.productRepo.split('/');
+        const productOwner = productSplitRepo[0];
+        const productRepo = productSplitRepo[1];
+        const resin = resinSdk();
+
+        let decodedToken: any;
+        let owner = '';
+        let repo = '';
+        let deployDetails: DeploymentDetails;
+
+        // Read the headers, validate the bearer token with the SDK.
+        const tokenMatch = headerToken.match(/^token (.*)$/i);
+        if (!tokenMatch) {
+            res.sendStatus(400);
+            return;
+        }
+
+        const token = tokenMatch[1];
+        resin.auth.loginWithToken(token).then(() => {
+            try {
+                decodedToken = jwtDecode(token);
+            } catch(_err) {
+                throw new Error('Cannot decode token into JWT object');
+            }
+
+            if (!_.includes(decodedToken.permissions, 'admin.home')) {
+                // Ensure it's a 401 so anyone without rights doesn't know it exists.
+                throw new HTTPError(401, 'Invalid access rights');
+            }
+
+            // Get the right environment.
+            const envRepo = this.environments[environment];
+            if (!envRepo) {
+                throw new HTTPError(404, 'Passed environment does not exist');
+            }
+
+            // Prep for the environment PR.
+            const splitRepo = envRepo.split('/');
+            owner = splitRepo[0];
+            repo = splitRepo[1];
+
+            // Ensure that the version of the keyframe specified actually exists.
+            return this.dispatchToEmitter(this.githubEmitterName, {
+                data: {
+                    owner: productOwner,
+                    repo: productRepo,
+                    path: KeyframeFilename,
+                    ref: `refs/tags/${version}`
+                },
+                method: this.githubApi.repos.getContent
+            });
+        }).then((keyframeFile: GithubApiTypes.Content) => {
+            // Github API docs state a blob will *always* be encoded base64...
+            if (keyframeFile.encoding !== 'base64') {
+                this.logger.log(LogLevel.WARN, `Keyframe file exists for ${productOwner}/${productRepo} but is not ` +
+                    `Base64 encoded! Aborting.`);
+                throw new HTTPError(500, 'Keyframe was not correctly encoded');
+            }
+
+            // We now go ahead and:
+            // 1. Create a new branch for this, create it from the version passed
+            // 2. Commit the keyframe to that branch
+            // 3. Open a new PR pointing to that branch. Any relevant reviewers can be set (when it works) from
+            //    a `.procbot.yml` config in the env repo.
+
+            // Create a new branch and commit the keyframe to it.
+            deployDetails = {
+                keyframe: keyframeFile,
+                username: decodedToken.username,
```

Comment:
Hmm, I'm not convinced by the username renaming, because the token obviously comes from the SDK, and therefore is a Resin user implicitly?
And no, no pinging occurs. And even if it did, they wouldn't be able to see it. :)